src/lib.rs

```diff
@@ -91,26 +85,36 @@ impl<T: Default> TinyVec<T> {
         data[*len] = val;
         *len += 1;
       }
-      Payload::Heap(ref mut vec) => {
-        vec.push(val)
-      }
+      Payload::Heap(ref mut vec) => vec.push(val),
     }
   }
 
   pub fn pop(&mut self) -> Option<T> {
     match &mut self.0 {
-      Payload::Inline { len: 0, .. } => {
-        None
-      }
+      Payload::Inline { len: 0, .. } => None,
       Payload::Inline { len, data } => {
         debug_assert!(*len > 0, "pop: illegal len: {}", len);
         let out = replace(&mut data[*len - 1], T::default());
         *len -= 1;
         Some(out)
       }
-      Payload::Heap(ref mut vec) => {
-        vec.pop()
+      Payload::Heap(ref mut vec) => vec.pop(),
+    }
+  }
+
+  pub fn truncate(&mut self, new_len: usize) {
+    match &mut self.0 {
+      Payload::Inline { len, data } => {
+        if needs_drop::<T>() {
+          while *len > new_len {
+            replace(&mut data[*len - 1], T::default());
+            *len -= 1;
```

Comment:
You could slice it once and then run an iterator to avoid bounds checks, but I'm not sure if that's going to be any faster.